docs/config/app_config.md
# mode testing and production and explain everything that is done during testing and production




# App Configuration Documentation

This document explains various configuration parameters used in the application.

## Initial Price Configuration

### `initial_price`

**Description:** Defines the initial price configuration needed to start the app.

**Parameters:**
- `symbol`: The trading pair symbol.
- `exchange`: The exchange where the symbol is traded.
- `instrument_type`: The type of instrument (e.g., perpetual).

## OKX Liquidation Settings Filter

### `okx_liquidations_filter`

**Description:** Settings related to OKX liquidation data filters.

**Parameters:**
- `instrument_types`: A list of instrument types to process (e.g., perpetual, future).
- `basecoin`: Basecoin to filter the data. If commented out, the symbols list will be used instead.
- `symbols`: A list of specific symbols to process. Uncomment and fill this list if filtering by symbols instead of basecoin.

## Fallback Symbols

### `fallback_symbols`

**Description:** A list of fallback symbols with their respective exchange and instrument type. If there is a stream that doesn’t have the price metric but price is needed to merge, this price will be used as a proxy. There must also be a stream of trades with this exact symbol.

**Parameters:**
- `symbol`: The trading pair symbol.
- `exchange`: The exchange where the symbol is traded.
- `instrument_type`: The type of instrument (e.g., perpetual).

## Data Merge Types

### `data_merge_types`

**Description:** A list of data merge types used for various data operations.

**Available Data Merge Types:**

- `spot_depth`: Depth of the spot market order book.
- `future_depth`: Depth of the perpetual market order book.
- `spot_trades`: Trades occurring in the spot market.
- `future_trades`: Trades occurring in the futures market.
- `option_trades`: Trades occurring in the options market.
- `future_oi_delta`: Changes in open interest for perpetual and futures contracts.
- `future_liquidations`: Liquidation events in the perpetual nad future markets.
- `option_oi`: Open interest in options contracts.
- `spot_canceled_depth`: Canceled orders in the spot market order book.
- `future_canceled_depth`: Canceled orders in the futures market order book.
- `spot_reinforced_depth`: Reinforced depth of the spot market order book.
- `future_reinforced_depth`: Reinforced depth of the futures market order book.

## Merge Intervals

### `merge_intervals_seconds`

**Description:** Specifies the intervals (in seconds) for merging data for different types.

**Parameters:**
- `spot`: Interval for spot data.
- `future`: Interval for future data.
- `option`: Interval for option data.




## Option Data Aggregation Settings

### `option_data_aggregation`
- **Description**: Settings for aggregating option open_interest data based on price ranges and expiry windows.
- **Parameters**:
  - `price_ranges`: A list of price values to create ranges for aggregation. The list can start from 0 but it is not necessary. It will create ranges of prices to merge on, such as -inf to x, x to y, and so on.
  - `expiry_windows`: A list of expiry values to create windows for aggregation. This list must contain 0 and does not contain negative values.
  - `fallback_symbols`: A dictionary containing symbols for index prices of options.
- **Example**:
    ```yaml
    option_data_aggregation:
      price_ranges: [0, 50, 100, 150, 200]  # Aggregates option open_interest data based on price ranges
      expiry_windows: [0, 7, 14, 30, 60]  # Aggregates option open_interest data based on expiry windows
      fallback_symbols:
        BTCUSD:
          index_symbol: BTCUSD_INDEX
        ETHUSD:
          index_symbol: ETHUSD_INDEX
    ```




option_data_aggregation
This section defines settings for aggregating option open interest data based on price ranges and expiry intervals.

price_buckets: A list of price values to create ranges (buckets) for aggregation. The list can start from 0 but it is not necessary. It will create ranges of prices to merge on, such as -inf to x, x to y, and so on. Can be float
expiry_intervals: A list of expiry values to create intervals for aggregation. This list must contain 0 and does not contain negative values.
fallback_symbols: A dictionary containing symbols for index prices of options. Only integers
This structure uses the new variable names price_buckets and expiry_intervals to clearly describe the purpose of the data they represent.





if one level size, then same for every instrument type
else not the same
spot_perpetual_future_data_aggregation:
  price_level_size: 20
  # price_level_size :
  #   spot : 20
  #   future : 20
  #   perpetual : 20


merge_interval: 60
# merge_interval:
#   spot: 60
#   perpetual: 60
#   future: 60
#   option: 60


# It can be tricky to merge futures OI, we need to balance it as the 20% difference of futures price is ok.

merge_types:
  spot:
    - depth
    - trades
    - canceled_depth
    - reinforced_depth
  perpetual_future:
    - depth
    - trades
    - oi
    - liquidations
    - canceled_depth
    - reinforced_depth
  option:
    - oi
    - depth
    - future
    - perpetual

  future:
    - depth
    - trades
    - oi
    - liquidations
    - canceled_depth
    - reinforced_depth

  perpetual:
    - depth
    - trades
    - oi
    - liquidations
    - canceled_depth
    - reinforced_depth

There can not be both perpetual_future and perpetual/future. THese are mutually exclusive. Threw an error if you see this

# Doesnt make sense to merge spot with anything just like options. Futures and perpetual can be merged

# None
# or daily_volume
# or open_interest
# options do not have weight factor

# merge_interval:
#   spot: 60
#   perpetual: 60
#   future: 60
#   option: 60



merge_weighting_factor: 
  spot:
    - depth: None 
    - trades: None 
  perpetual:
    - depth: None  
    - trades: None 
    - oi: None     
    - liquidations: None 
  future:
    - depth: None  
    - trades: None 
    - oi: None     
    - liquidations: None     
  perpetual_future:
    - depth: None  
    - trades: None
    - oi: None     
    - liquidations: None    



# If you really enjoy pain then do this

  <!-- # price_level_size :
  #   spot : 20
  #   future : 20
  #   perpetual : 20
  #   perpetual_future : 20  # IF you merge future_perpetual instruments

# calculate_canceled_depth:
#   spot: True
#   # perpetual: True
#   # future: True 
#   # perpetual_future: True
#   # option : True
# calculate_reinforced_depth:
#   spot: True
#   # perpetual: True
#   # future: True 
#   # perpetual_future: True
#   # option : True -->

docs/config/kafka_config.md
# Kafka Configuration Parameters

This document explains various Kafka configuration parameters with typical settings and links to Kafka documentation for further reading.

## Message Parameters

### `message.max.bytes`
- **Description**: Maximum size of a message that the broker will accept.
- **Typical Setting**: Larger than your largest expected message size (e.g., 1MB if expecting large messages).
- **Documentation**: [Kafka Documentation](https://kafka.apache.org/documentation/#brokerconfigs_message.max.bytes)

### `replica.fetch.max.bytes`
- **Description**: Maximum number of bytes fetch requests are allowed to fetch from a single partition.
- **Typical Setting**: Same or slightly larger than `message.max.bytes`.
- **Documentation**: [Kafka Documentation](https://kafka.apache.org/documentation/#brokerconfigs_replica.fetch.max.bytes)

### `fetch.message.max.bytes`
- **Description**: Maximum number of bytes fetch requests are allowed to fetch from a single partition. This must be at least as large as the largest message you want to fetch.
- **Typical Setting**: Match with `message.max.bytes`.
- **Documentation**: [Kafka Documentation](https://kafka.apache.org/documentation/#consumerconfigs_fetch.message.max.bytes)

## Thread Parameters

### `num.network.threads`
- **Description**: The number of threads that the server uses for processing requests, which are received over the network.
- **Typical Setting**: Depends on your server's hardware, but starting with a value equal to the number of cores is a good approach.
- **Documentation**: [Kafka Documentation](https://kafka.apache.org/documentation/#brokerconfigs_num.network.threads)

### `num.io.threads`
- **Description**: The number of threads that perform disk I/O operations in the broker.
- **Typical Setting**: Equal to or greater than the number of storage devices.
- **Documentation**: [Kafka Documentation](https://kafka.apache.org/documentation/#brokerconfigs_num.io.threads)

## Socket Parameters

### `socket.receive.buffer.bytes`
- **Description**: The receive buffer (`SO_RCVBUF`) used by the network sockets.
- **Typical Setting**: Increase if noticing network issues, typical values are 102400 (100KB) to 204800 (200KB).
- **Documentation**: [Kafka Documentation](https://kafka.apache.org/documentation/#brokerconfigs_socket.receive.buffer.bytes)

### `socket.send.buffer.bytes`
- **Description**: The send buffer (`SO_SNDBUF`) used by the network sockets.
- **Typical Setting**: Same as `socket.receive.buffer.bytes`.
- **Documentation**: [Kafka Documentation](https://kafka.apache.org/documentation/#brokerconfigs_socket.send.buffer.bytes)

## Queue Parameters

### `queued.max.request.bytes`
- **Description**: The maximum number of bytes in the request queue. A large request queue can handle a high volume of requests but may use more memory.
- **Typical Setting**: Adjust based on memory availability and load.
- **Documentation**: [Kafka Documentation](https://kafka.apache.org/documentation/#brokerconfigs_queued.max.request.bytes)

## Producer Parameters

### `max.request.size`
- **Description**: Maximum size of a request that the producer can send.
- **Typical Setting**: Match or exceed `message.max.bytes`.
- **Documentation**: [Kafka Documentation](https://kafka.apache.org/documentation/#producerconfigs_max.request.size)

### `buffer.memory`
- **Description**: Total bytes of memory the producer can use to buffer records waiting to be sent to the server.
- **Typical Setting**: Higher values can handle more records held during high traffic bursts.
- **Documentation**: [Kafka Documentation](https://kafka.apache.org/documentation/#producerconfigs_buffer.memory)

### `compression.type`
- **Description**: Specify the compression type for all data generated by the producer. Options include none, gzip, snappy, lz4, or zstd.
- **Typical Setting**: gzip or snappy for a balance between compression rate and CPU usage.
- **Documentation**: [Kafka Documentation](https://kafka.apache.org/documentation/#producerconfigs_compression.type)

### `linger.ms`
- **Description**: The producer will wait this long before sending records if the batch isn't filled.
- **Typical Setting**: Increases batch size and throughput at the cost of latency.
- **Documentation**: [Kafka Documentation](https://kafka.apache.org/documentation/#producerconfigs_linger.ms)

### `batch.size`
- **Description**: Maximum number of bytes that will be included in a batch of messages.
- **Typical Setting**: Increase to maximize throughput, especially if you are sending large messages.
- **Documentation**: [Kafka Documentation](https://kafka.apache.org/documentation/#producerconfigs_batch.size)

### `acks`
- **Description**: The number of acknowledgments the producer requires from brokers.
- **Typical Setting**: all for ensuring full durability of records.
- **Documentation**: [Kafka Documentation](https://kafka.apache.org/documentation/#producerconfigs_acks)

### `retry.backoff.ms`
- **Description**: The amount of time to wait before attempting to retry a failed produce request.
- **Typical Setting**: Adjust based on your application's tolerance for retries.
- **Documentation**: [Kafka Documentation](https://kafka.apache.org/documentation/#producerconfigs_retry.backoff.ms)

config.py
# config.py

coinbase_secret = '-----BEGIN EC PRIVATE KEY-----\nMHcCAQEEIDOVctxJpAI/hHtbUN9VrHej4bWPRuT9um9FoBlTgiyaoAoGCCqGSM49\nAwEHoUQDQgAEJt8JWIh8CHm045POImBF0ZvVuX5FbQjIDhIT82hE5r1+vb8cSQ3M\nfEjriBy1/ZD3EywPNxyGe6nO/Wsq0M8hXQ==\n-----END EC PRIVATE KEY-----\n'
coinbase_api = 'organizations/b6a02fc1-cbb0-4658-8bb2-702437518d70/apiKeys/697a8516-f2e2-4ec9-a593-464338d96f21'
kucoin_api = "65d92cc0291aa2000118b67b"
kucoin_secret = "3d449464-ab5e-4415-9950-ae31648fe90c"
kucoin_pass = "sN038-(!UK}4"
elastic_password = "P0H3MsMaUKy7=6jOi-Jx"

# password: HNtNChGd-2Id-9O_XXg_
#           P0H3MsMaUKy7=6jOi-Jx
# certificate SHA-256 fingerprint: e13b8081995a509320d2882434ec223657befe375fb81f7753565e030afecf9b
# kibana enrolement token: eyJ2ZXIiOiI4LjE0LjAiLCJhZHIiOlsiMTkyLjE2OC4xLjI1Mjo5MjAwIl0sImZnciI6ImUxM2I4MDgxOTk1YTUwOTMyMGQyODgyNDM0ZWMyMjM2NTdiZWZlMzc1ZmI4MWY3NzUzNTY1ZTAzMGFmZWNmOWIiLCJrZXkiOiIzUnlKYVpBQjJaU1RKQktrM2FHZjp0NzdGQ1Fta1RpYU9zR3NqTVZlVi1nIn0=
# nodes_enrolement_token = eyJ2ZXIiOiI4LjE0LjAiLCJhZHIiOlsiMTkyLjE2OC4xLjI1Mjo5MjAwIl0sImZnciI6ImUxM2I4MDgxOTk1YTUwOTMyMGQyODgyNDM0ZWMyMjM2NTdiZWZlMzc1ZmI4MWY3NzUzNTY1ZTAzMGFmZWNmOWIiLCJrZXkiOiIzaHlQYVpBQjJaU1RKQktra0tHajozUXJodkRQZ1Q3LVpSSVZUd05uM2ZnIn0=


kafka_configs = {
    "message.max.bytes": 0,  # Description: Maximum size of a message that the broker will accept.
                             # Typical Setting: Larger than your largest expected message size (e.g., 1MB if expecting large messages).

    "replica.fetch.max.bytes": 0,  # Description: Maximum number of bytes fetch requests are allowed to fetch from a single partition.
                                   # Typical Setting: Same or slightly larger than message.max.bytes.
    
    "fetch.message.max.bytes": 0,  # Description: Maximum number of bytes fetch requests are allowed to fetch from a single partition.
                                   # This must be at least as large as the largest message you want to fetch.
                                   # Typical Setting: Match with message.max.bytes.
    
    "num.network.threads": 0,  # Description: The number of threads that the server uses for processing requests, which are received over the network.
                               # Typical Setting: Depends on your server's hardware, but starting with a value equal to the number of cores is a good approach.
    
    "num.io.threads": 0,  # Description: The number of threads that perform disk I/O operations in the broker.
                          # Typical Setting: Equal to or greater than the number of storage devices.
    
    "socket.receive.buffer.bytes": 0,  # Description: The receive buffer (SO_RCVBUF) used by the network sockets.
                                       # Typical Setting: Increase if noticing network issues, typical values are 102400 (100KB) to 204800 (200KB).
    
    "socket.send.buffer.bytes": 0,  # Description: The send buffer (SO_SNDBUF) used by the network sockets.
                                    # Typical Setting: Same as socket.receive.buffer.bytes.
    
    "queued.max.request.bytes": 0,  # Description: The maximum number of bytes in the request queue.
                                    # A large request queue can handle a high volume of requests but may use more memory.
                                    # Typical Setting: Adjust based on memory availability and load.
    
    # Kafka Producer Configuration Parameters
    "max.request.size": 0,  # Description: Maximum size of a request that the producer can send.
                            # Typical Setting: Match or exceed message.max.bytes.
    
    "buffer.memory": 0,  # Description: Total bytes of memory the producer can use to buffer records waiting to be sent to the server.
                         # Typical Setting: Higher values can handle more records held during high traffic bursts.
    
    "compression.type": "none",  # Description: Specify the compression type for all data generated by the producer.
                                 # Options include none, gzip, snappy, lz4, or zstd.
                                 # Typical Setting: gzip or snappy for a balance between compression rate and CPU usage.
    
    "linger.ms": 0,  # Description: The producer will wait this long before sending records if the batch isn't filled.
                     # Typical Setting: Increases batch size and throughput at the cost of latency.
    
    "batch.size": 0,  # Description: Maximum number of bytes that will be included in a batch of messages.
                      # Typical Setting: Increase to maximize throughput, especially if you are sending large messages.
    
    "acks": "all",  # Description: The number of acknowledgments the producer requires from brokers.
                    # Typical Setting: all for ensuring full durability of records.
    
    "retry.backoff.ms": 0,  # Description: The amount of time to wait before attempting to retry a failed produce request.
                            # Typical Setting: Adjust based on your application's tolerance for retries.
}

default_price_value = 70000

# okx are SWAP for pepretual, FUTURE for futures and OPTION for options
okx_liquidations_inst_types = ["perpetual", "future"] #Okx only provides a unique stream for all liquidations. Those symbol found here will be processed others omited

#
okx_liquidations_symbols = {
        "process_type" : "basecoin", # basecoin, symbols
        "basecoin" : "BTC",  # Processes data of symbols with this basecoin
        "symbols" : []       # Processes only this symbols, if not empty. If empty then all
    } 


backup_symbols = { # you may cross symbols from different exchanges
    "binance" : "BTCUSDT@perpetual@binance",
    "bybit" : "BTCUSDT@perpetual@bybit",
    "okx" : "BTC-USDT-SWAP@perpetual@okx",
    "deribit" : "BTC-PERPETUAL@perpetual@deribit",
    "bitget" : "BTCUSDT@perpetual@bitget",
    "bingx" : "BTC-USDT@perpetual@bingx",
    "htx" : "BTCUSDT@perpetual@binance",
    "kucoin" : "XBTUSDTM@perpetual@kucoin",
    "mexc" : "BTC_USDT@perpetual@mexc",
    "gateio" : "BTC_USDT@perpetual@gateio",
}

merge_types = [
    "depth_spot", 
    "depth_future", 
    "trades_spot", 
    "trades_future", 
    "trades_option", 
    "oi_delta_future", 
    "liquidations_future",
    "oi_options", 
    "cdepth_spot", 
    "cdepth_future", 
    "rdepth_spot", 
    "rdepth_future"
    ]


message_processor_fail_threshold = 0.1


# MULTIPLIER CONFIGURATION

## BINANCE ###

# Contract values for Binance linear contracts (stablecoin/fiat-margined contracts) are standardized to 1 BTC,
# so no multiplication is needed.

# However, for coin-margined contracts, the contract values vary:
# - BTC contracts are worth 100 USD
# - Altcoin (other) contracts are worth 10 USD
# - If binance changes the contract values, this config file will need to be updated. simply add the basecoin and the following multiplier (ex : {"RBNB : 12"})

# This same principle applies to options contracts. But, options are worth in native coin.
binance_inverse_multipliers = {
            "BTC" : 100,
            "other" : 10
        }
binance_option_multipliers = {
            "BTC" : 1,
            "ETH" : 1,
            "BNB" : 1,
            "XRP" : 100,
            "DOGE" : 1000,
        }


### BYBIT ###

# Bybit USDT and USDC contracts worth 1 BTC, no need to multiply. 
# But, coin-marginated (inverse) contracts worth 1 usd. No need to config
# Options also worth 1 btc

### OKX ###
# okx is simmilar to binance though a bit different.
# if there is a lambda function instead of multiplier, its because the multiplier is worth in native coin rather than the stablecoin
okx_future_multipliers = {
    "USDT" : {
        "BTC" : lambda amount : amount * 0.01, 
        "ETH" : lambda amount : amount * 0.1, 
        "XLM" : 100,
        "TRX" : 1000,
        "other" : 10,
    },
    "USDC" : {
        "BTC" : lambda amount : amount * 0.0001,
        "ETH" : lambda amount : amount * 0.001,
        "XLM" : 100,
        "TRX" : 1000,
        "other" : 10, 
    },
    "USD" : {
        "XLM" : 100,
        "BTC" : 100,
        "TRX" : 1000,
        "other" : 10
    }
}
# options are worth in native coin 
okx_option_multipliers = {
            "BTC" : 0.01,
            "ETH " : 0.1,
            "LTC" : 1,
            "LINK" : 10,
            "DOT" : 10,
            "SOL" : 10,
            "ADA" : 100,
        }


### DERIBIT ###

deribit_future_multipliers = {
    "perpetual" : {
        "BTC-PERPETUAL" : lambda amount, price : amount  / price / 10,
        "ETH-PERPETUAL" : lambda amount, price : amount  / price,
        "ADA_USDC-PERPETUAL" : lambda amount, *args, **kwargs: amount * 10,
        "ALGO_USDC-PERPETUAL" : lambda amount, *args, **kwargs: amount * 10,
        "AVAX_USDC-PERPETUAL" : lambda amount, *args, **kwargs: amount * 0.1,
        "BCH_USDC-PERPETUAL" : lambda amount, *args, **kwargs: amount * 0.02,
        "BTC_USDC-PERPETUAL" : lambda amount, *args, **kwargs: amount * 0.001,
        "ETH_USDC-PERPETUAL" : lambda amount, *args, **kwargs: amount * 0.01,
        "LINK_USDC-PERPETUAL" : lambda amount, *args, **kwargs: amount,
        "LTC_USDC-PERPETUAL" : lambda amount, *args, **kwargs: amount * 0.1,
        "MATIC_USDC-PERPETUAL" : lambda amount, *args, **kwargs: amount * 10,
        "LINK_USDT-PERPETUAL" : lambda amount, *args, **kwargs: amount,
        "LTC_USDT-PERPETUAL" : lambda amount, *args, **kwargs: amount * 0.1,
        "NEAR_USDC-PERPETUAL" : lambda amount, *args, **kwargs: amount,
        "SOL_USDC-PERPETUAL" : lambda amount, *args, **kwargs: amount * 0.1,
        "TRX_USDC-PERPETUAL" : lambda amount, *args, **kwargs: amount * 100,
        "UNI_USDC-PERPETUAL" : lambda amount, *args, **kwargs: amount ,
        "XRP_USDC-PERPETUAL" : lambda amount, *args, **kwargs: amount * 10
    },
    "future" : {
    "BTC" : lambda amount, price : amount  / price / 10,
    "ETH" : lambda amount, price : amount  / price  
 }
}


deribit_option_multipliers = {
            "BTC" : 1,
            "ETH" : 1,
            "MATIC" : 1000,
            "SOL" : 10,
            "XRP" : 1000,
        }


### KUCOIN ###

# https://www.kucoin.com/pt/futures/contract/detail/ACEUSDTM
# Find you coin here and then add
kucoin_future_multipliers = {
    "XBTUSDTM" : 0.001,
    "XBTUSDTP" : 0.001,
    "XBTUSDM" : lambda amount, price: amount / price,
}


### MEXC ###
# https://www.mexc.com/support/articles/17827791509072
mexc_future_multipliers = {
    "BTC_USDT" : 0.0001,
    "BTC_USD" : lambda amount, price : amount * 100  / price
}

### GATEIO ###
# https://www.gate.io/futures_info_new/futures/usdt/BTC_USDT#baseinfo
gateio_future_multipliers = {
    "BTC_USDT" : 0.0001,
    "BTC_USD" : lambda amount, price : amount / price
}

gateio_option_multipliers = {
    "BTC_USDT" : 0.01,
    "ETH_USDT" : 0.01,
    "DOGE_USDT" : 1000,
}


fiats = [
    "usd", 
    "eur", 
    "rub", 
    "try", 
    "uah", 
    "kzt", 
    "inr", 
    "gbp", 
    "krw", 
    "aud", 
    "chf", 
    "czk", 
    "dkk", 
    "nok", 
    "nzd", 
    "pln", 
    "sek", 
    "zar", 
    "huf", 
    "ils"
]
stablecoins = [
    "usdt", 
    "usdc", 
    "busd", 
    "dai", 
    "tusd",
    "fdusd"
    ]